system/core/macros.php

```diff
@@ -0,0 +1,185 @@
+<?php
+
+/**
+ * Superclass for all macros.
+ */
+abstract class HyphaMacro {
+	/**
+	 * The RequestContext this macro is evaluated in.
+	 * @var RequestContext
+	 */
+	protected $O_O;
+
+	/**
+	 * The <macro> tag this object represents and should replace.
+	 *
+	 * @var HyphaDomElement
+	 */
+	protected $macro_tag;
+
+	/**
+	 * The document that contains $macro_tag.
+	 *
+	 * @var HTMLDocument
+	 */
+	protected $doc;
+
+	/**
+	 * The context this macro is evaluated in.
+	 *
+	 * @var array
+	 */
+	protected $macro_context;
+
+	public function __construct(RequestContext $O_O, HyphaDomElement $macro_tag, array $macro_context) {
+		$this->O_O = $O_O;
+		$this->macro_tag = $macro_tag;
+		$this->doc = $macro_tag->document();
+		$this->macro_context = $macro_context;
+	}
+
+	/**
+	 * Invoke this macro.
+	 *
+	 * This method should either return a HyphaDomElement or a
+	 * NodeList, which will then replace for the macro
+	 * element.
+	 *
+	 * Alternatively, if more complex replacement is needed, this
+	 * method can also return false, in which case the macro element
+	 * is simply deleted (assuming this method has then already
+	 * inserted new content elsewhere).
+	 */
+	abstract public function invoke();
+
+	/**
+	 * Helper that copies common attributes from the macro_tag to
+	 * the given element.
+	 *
+	 * Currently copies `id` and `class`, more could be added in the
+	 * future. The `class` attribute is merged, other attributes are
+	 * replaced.
+	 */
+	protected function copyAttributesTo(HyphaDomElement $to) {
+		$from = $this->macro_tag;
+		if ($from->hasAttribute('id'))
+			$to->setAttribute('id', $from->getAttribute('id'));
+		if ($from->hasAttribute('class'))
+			$to->addClass($from->getAttribute('class'));
+	}
+
+	/**
+	 * Gets the attribute with the given name from the macro tag and
+	 * converts it to DateTime object using the given format (as
+	 * defined in DateTime::createFromFormat). If the attribute is
+	 * not present, the given default is returned. If the value is
+	 * invalid, an UnexpectedValueException is raised, using the
+	 * given errorFormat to indicate the expected format to the
+	 * user.
+	 *
+	 * @param string $name The name of the attribute to get
+	 * @return DateTime | null
+	 */
+	protected function getDateTimeAttribute($name, $format, $errorFormat, $default = null) {
```

Comment:
I'm a bit surprised to see this function here. I wouldn't expect it to be part of the base class of a macro, seems a bit specific. But perhaps I'm missing something?

Reply:
Well, I could imagine other macros that also need date/time parameters, so this is just a helper function to access such attributes/arguments. I also anticipate other helper functions for parsing other types of arguments, but I haven't had the need yet I guess.

Maybe it would make sense to move the actual parsing code to a more general place, and then here just call that?

Reply:
I'm going to leave this for now and merge this. We can always move this code later if that turns out better (e.g. when we add a few more macros).

Reply:
Cool :)